programmers/실패율.js

```diff
@@ -0,0 +1,15 @@
+function solution(N, stages) {
+  const challengingStageNumbers = Array.from(new Array(N + 1), (_, k) => k + 1).map((stageNumber) => {
+    return stages.filter((stage) => stage === stageNumber).length;
+  });
+
+  const failureRatios = challengingStageNumbers.map((challengingStageNumber, index) => {
+    const challengeStageNumber = challengingStageNumbers.slice(index).reduce((acc, challengeNum) => acc + challengeNum, 0);
+    return challengingStageNumber / challengeStageNumber;
+  }).slice(0, -1);
+
+  const failureRatioRank = failureRatios.map((failureRatio, index) => [index + 1, failureRatio])
+  .sort((a, b) => b[1] - a[1]).map((val) => val[0]);
+  
+  return failureRatioRank;
+}
```

Comment:
줄바꿈을 좀 더 신경써주시길 부탁드릴게요.
index 변수명을 고집하는 이유가 있나요?
idx나 i 같이 짧은 이름도 있는데 왜 굳이 index를 사용하시는건가요?

```suggestion
function solution(N, stages) {
  const challengersByStage = Array.from(
    new Array(N + 1),
    (_, k) => k + 1
  ).map((stageNumber) => {
    return stages.filter((stage) => stage === stageNumber).length;
  });

  const failureRatios = challengersByStage
    .map((challengingStageNumber, index) => {
      const challengeStageNumber = challengersByStage
        .slice(index)
        .reduce((acc, challengeNum) => acc + challengeNum, 0);

      return challengingStageNumber / challengeStageNumber;
    })
    .slice(0, -1);

  const failureRatioRank = failureRatios
    .map((failureRatio, index) => [index + 1, failureRatio])
    .sort((a, b) => b[1] - a[1])
    .map((val) => val[0]);

  return failureRatioRank;
}
```

줄바꿈만 신경써도 훨씬 좋은 코드가 될 수 있어요.

Reply:
아래 풀이도 쓰윽 함 봐보세요!
메서드를 줄줄이 잇는 것은 편리하지만 가독성을 떨어뜨리는 문제점이 있어요.
따라서 줄바꿈을 해주며 깔끔하게 정리를 무조건 해주셔야 해요.

```js
function solution(N, stages) {
    const stageCnt = {};
    const failRate = {};
    
    // 스테이지 갯수 카운팅
    stages.forEach((n) => {
        if (stageCnt[n]) stageCnt[n]++;
        else stageCnt[n] = 1;
    });
    
    // 도전자 수
    let challenger = stages.length; 
    
    // 실패율 구하기
    for(let n = 1; n < N + 1; n++) {
        failRate[n] = stageCnt[n] ? stageCnt[n] / challenger : 0; // 실패율
        failRate[n] = challenger === 0 ? 0 : failRate[n]; // 도전자가 없으면 실패율 0으로 초기화
        challenger -= stageCnt[n] ? stageCnt[n] : 0; // 스테이지별 도전자수 차감
    }
    
    // 실패율에 따라 정렬하기
    return Object.entries(failRate)
        .sort((a, b) => {
            if(a[1] === b[1]) return a[1] - b[1];
            else return b[1] - a[1];
        })
        .map((el) => parseInt(el[0]));
}
```